test/linalg/tridiag.jl

```diff
@@ -228,118 +16,180 @@ end
+                @test_throws ArgumentError A[1, 3]   = 1 # test assignment off the main/sub/super diagonal
+                if mat_type == Tridiagonal
+                    @test (A[3, 3] = A[3, 3]; A == fA) # test assignment on the main diagonal
+                    @test (A[3, 2] = A[3, 2]; A == fA) # test assignment on the subdiagonal
+                    @test (A[2, 3] = A[2, 3]; A == fA) # test assignment on the superdiagonal
+                    @test ((A[1, 3] = 0) == 0; A == fA) # test zero assignment off the main/sub/super diagonal
+                else
+                    @test ((A[3, 3] = A[3, 3]) == A[3, 3]; A == fA) # test assignment on the main diagonal
+                    @test_throws ArgumentError A[3, 2] = 1 # test assignment on the subdiagonal
+                    @test_throws ArgumentError A[2, 3] = 1 # test assignment on the superdiagonal
+                end
             end
             @testset "Diagonal extraction" begin
-                @test diag(A,1) == b
-                @test diag(A,-1) == b
-                @test diag(A,0) == a
-                @test diag(A,n-1) == zeros(elty,1)
-                @test_throws ArgumentError diag(A,n+1)
+                @test diag(A, 1) == (mat_type == Tridiagonal ? du : dl)
+                @test diag(A, -1) == dl
+                @test diag(A, 0) == d
+                @test diag(A, n - 1) == zeros(elty, 1)
+                @test_throws ArgumentError diag(A, n + 1)
             end
             @testset "Idempotent tests" begin
                 for func in (conj, transpose, adjoint)
                     @test func(func(A)) == A
                 end
             end
-            @testset "Simple unary functions" begin
-                for func in (det, inv)
-                    @test func(A) ≈ func(fA) atol=n^2*sqrt(eps(relty))
+            if elty != Int
+                @testset "Simple unary functions" begin
+                    for func in (det, inv)
+                        @test func(A) ≈ func(fA) atol=n^2*sqrt(eps(real(one(elty))))
+                    end
                 end
             end
-
+            ds = mat_type == Tridiagonal ? (dl, d, du) : (d, dl)
+            for f in (real, imag)
+                @test f(A) == mat_type(map(f, ds)...)
+            end
             if elty <: Real
-                @testset "Rounding to Ints" begin
-                    @test round.(Int,A) == round.(Int,fA)
-                    @test isa(round.(Int,A), SymTridiagonal)
-                    @test trunc.(Int,A) == trunc.(Int,fA)
-                    @test isa(trunc.(Int,A), SymTridiagonal)
-                    @test ceil.(Int,A) == ceil.(Int,fA)
-                    @test isa(ceil.(Int,A), SymTridiagonal)
-                    @test floor.(Int,A) == floor.(Int,fA)
-                    @test isa(floor.(Int,A), SymTridiagonal)
+                for f in (round, trunc, floor, ceil)
+                    fds = [f.(d) for d in ds]
+                    @test f.(A) == mat_type(fds...)
```

Comment:
Perhaps remove the `isa` tests here and do `@test (f.(A))::mat_type == mat_type(fds...)` for even more code deletion.